xbmc/addons/binary-addons/BinaryAddonBase.cpp

```diff
@@ -0,0 +1,383 @@
+
+      /*
+       * Parse addon.xml "<assets>...</assets>"
+       */
+      const TiXmlElement* element = child->FirstChildElement("assets");
+      if (element)
+      {
+        for (const TiXmlElement* elementsAssets = element->FirstChildElement(); elementsAssets != nullptr; elementsAssets = elementsAssets->NextSiblingElement())
+        {
+          std::string value = elementsAssets->Value();
+          if (value == "icon")
+          {
+            if (elementsAssets->GetText() != nullptr)
+              m_icon = elementsAssets->GetText();
+          }
+          else if (value == "fanart")
+          {
+            if (elementsAssets->GetText() != nullptr)
+              m_art[value] = elementsAssets->GetText();
+          }
+          else if (value == "banner")
+          {
+            if (elementsAssets->GetText() != nullptr)
+              m_art[value] = elementsAssets->GetText();
+          }
+          else if (value == "clearlogo")
+          {
+            if (elementsAssets->GetText() != nullptr)
+              m_art[value] = elementsAssets->GetText();
+          }
+        }
+      }
+
+      /* Parse addon.xml "<noicon">...</noicon>" */
+      if (m_icon.empty())
+      {
+        element = child->FirstChildElement("noicon");
+        m_icon = (element && strcmp(element->GetText() , "true") == 0) ? "" : "icon.png";
+      }
+
+      /* Parse addon.xml "<nofanart">...</nofanart>" */
+      if (m_art.empty())
+      {
+        element = child->FirstChildElement("nofanart");
+        if (!element || strcmp(element->GetText(), "true") != 0)
+          m_art["fanart"] = URIUtils::AddFileToFolder(addonPath, "fanart.jpg");
+      }
+
+      /* Parse addon.xml "<news lang="..">...</news>" */
+      const char* strChangelog = nullptr;
+      element = child->FirstChildElement("news");
+      while (element)
+      {
+        const char *lang = element->Attribute("lang");
+        if (lang != nullptr && g_langInfo.GetLocale().Matches(lang))
+        {
+          strChangelog = element->GetText();
+          break;
+        }
+        else if (lang == nullptr || strcmp(lang, "en") == 0 || strcmp(lang, "en_GB") == 0)
+        {
+          strChangelog = element->GetText();
+        }
+
+        element = element->NextSiblingElement("news");
+      }
+      m_changelog = strChangelog ? strChangelog : "";
+    }
+    else
+    {
+      TYPE type = CAddonInfo::TranslateType(point);
+      if (type == ADDON_UNKNOWN || type >= ADDON_MAX)
+      {
+        CLog::Log(LOGERROR, "CBinaryAddonBase: file '%s' doesn't contain a valid add-on type name (%s)", addonPath.c_str(), point.c_str());
+        return false;
+      }
+
+      m_types.push_back(CBinaryAddonType(type, this, child));
+    }
+  }
+
+  /*
+   * If nothing is defined in addon.xml set this as unknown to have minimum one
+   * instance type present.
+   */
+  if (m_types.empty())
+  {
+    CBinaryAddonType addonType(ADDON_UNKNOWN, this, nullptr);
+    m_types.push_back(addonType);
```

Comment:
no need for the variable here. could just read <code>m_types.emplace_back(CBinaryAddonType(ADDON_UNKNOWN, this, nullptr));</code>